src/dispatch/batch_builder.rs

```diff
@@ -0,0 +1,496 @@
+///         self.iterations < 3
+///     }
+/// }
+///
+/// impl<'a> System<'a> for TestBatchControllerSystem {
+///
+///     type SystemData = ();
+///
+///     fn run(&mut self, (): Self::SystemData){
+///
+///         self.iterations += 1;
+///     }
+///
+/// }
+///
+/// ```
+///
+/// Use the function `BatchBuilder::with_batch_controller` or `BatchBuilder::set_batch_controller`
+/// to set your BatchControllerSystem.
+pub trait BatchController {
+    /// This function is called at the start of the BatchSystem execution.
+    fn prepare_dispatching(&mut self);
+
+    /// This control the internal dispatching of the sub systems.
+    ///
+    /// The following pseudo code show how it works.
+    ///
+    /// >loop{
+    /// >    if batch_controller.want_to_dispatch() == false {
+    /// >        break
+    /// >    }
+    /// >
+    /// >    substep Dispatching
+    /// >}
+    ///
+    fn want_to_dispatch(&mut self) -> bool;
+}
+
+pub trait BatchControllerSystem<'a>: BatchController + RunNow<'a> {}
+
+impl<'a, T> BatchControllerSystem<'a> for T where T: BatchController + RunNow<'a> {}
+
+/// This executes only 1 time the BatchSystem.
+/// Use this controller doesn't make much
+/// sense for this reason an error message will be print.
+#[derive(Default)]
+pub struct DefaultBatchControllerSystem {
+    is_first_execution: bool,
+}
+
+impl BatchController for DefaultBatchControllerSystem {
+    fn prepare_dispatching(&mut self) {
+        self.is_first_execution = true;
+    }
+
+    fn want_to_dispatch(&mut self) -> bool {
+        self.is_first_execution
+    }
+}
+
+impl<'a> System<'a> for DefaultBatchControllerSystem {
+    type SystemData = ();
+
+    fn run(&mut self, (): Self::SystemData) {
+        self.is_first_execution = false;
+        eprintln!("The default batch controller is in use. Consider to use a normal System.")
+    }
+}
+
+/// The BatchBuilder is responsible for the creation of the Batch.
+/// It works similarly to the DispatcherBuilder, indeed is possible to set
+/// the systems and the barriers.
+/// The additional feature is that you can specify a BatchControllerSystem
+/// to control the execution of the Batch.
+///
+/// Despite the similarities, it does a different thing since it create a Batch.
+/// The Batch allow to define a group of Systems that can be dispatched
+/// multiple times within the same main dispatching process.
+/// For this reason the Batch dispatching is called SubDispatching.
+///
+/// Note that depending on the dependencies of the SubSystems the Batch
+/// can run in parallel with other Systems.
+/// In addition the Sub Systems can run in parallel within the Batch.
+pub struct BatchBuilder<'a> {
+    current_id: usize,
+    map: HashMap<String, SystemId>,
+    stages_builder: StagesBuilder<'a>,
+    read_resources: Vec<ResourceId>,
+    write_resources: Vec<ResourceId>,
+    running_time: f32,
+    system_count: u32,
+    batch_controller: Box<for<'c> BatchControllerSystem<'c> + Send + 'a>,
```

Comment:
I don't quite see why this should be controlled like this (it also seems like a special case, which shouldn't be in shred).

Instead, execution should be controlled by wrapping the `BatchSystem` (or, preferably, the `Dispatcher`, see my other comment).